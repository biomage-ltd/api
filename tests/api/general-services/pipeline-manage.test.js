const AWSMock = require('aws-sdk-mock');
const _ = require('lodash');
const AWS = require('../../../src/utils/requireAWS');

const createPipeline = require('../../../src/api/general-services/pipeline-manage');

jest.mock('uuid', () => ({
  v4: () => 'mock-uuid',
}));

describe('test for pipeline services', () => {
  afterEach(() => {
    AWSMock.restore('EKS');
    AWSMock.restore('StepFunctions');
  });


  const MockProcessingConfig = {
    Item: {
      processingConfig: {
        M: {
          doubletScores: {
            M: {
              filterSettings: {
                M: {
                  oneSetting: {
                    N: 1,
                  },
                },
              },
              oneSample: {
                M: {
                  filterSettings: {
                    M: {
                      oneSetting: {
                        N: 1,
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  };

  const MockSamples = {
    Item: {
      samples: {
        M: {
          ids: {
            L: [
              { S: 'sample-KO' },
            ],
          },
        },
      },
    },

  };

  const mockCluster = {
    cluster: {
      name: 'biomage-test',
      endpoint: 'https://test-endpoint.me/fgh',
      certificateAuthority: {
        data: 'AAAAAAAAAAA',
      },
    },
  };

  const processingConfigUpdate = [
    {
      name: 'doubletScores',
      body: {
        oneSample: {
          filterSettings: {
            oneSetting: 7,
          },
        },
        otherSample: {
          filterSettings: {
            oneSetting: 15,
          },
        },
      },
    },
  ];

  it('Create pipeline works', async () => {
    AWSMock.setSDKInstance(AWS);

    const describeClusterSpy = jest.fn((x) => x);
    AWSMock.mock('EKS', 'describeCluster', (params, callback) => {
      describeClusterSpy(params);
      callback(null, mockCluster);
    });

    const createStateMachineSpy = jest.fn(
      (stateMachineObject) => _.omit(stateMachineObject, ['definition', 'image']),
    );

    AWSMock.mock('StepFunctions', 'createStateMachine', (params, callback) => {
      createStateMachineSpy(params);
      callback(null, { stateMachineArn: 'test-machine' });
    });

    const startExecutionSpy = jest.fn((x) => x);
    AWSMock.mock('StepFunctions', 'startExecution', (params, callback) => {
      startExecutionSpy(params);
      callback(null, { executionArn: 'test-machine' });
    });

    const getProcessingConfigSpy = jest.fn((x) => x);
    const getSamplesSpy = jest.fn((x) => x);
    AWSMock.mock('DynamoDB', 'getItem', (params, callback) => {
      if (params.TableName.match('experiments')) {
        getProcessingConfigSpy(params);
        callback(null, MockProcessingConfig);
      } else if (params.TableName.match('samples')) {
        getSamplesSpy(params);
        callback(null, MockSamples);
      }
    });

    await createPipeline('testExperimentId', processingConfigUpdate);
    expect(describeClusterSpy).toMatchSnapshot();

    expect(createStateMachineSpy.mock.results).toMatchSnapshot();

    expect(getProcessingConfigSpy).toHaveBeenCalled();
    expect(getSamplesSpy).toHaveBeenCalled();

    expect(startExecutionSpy).toHaveBeenCalled();
    expect(startExecutionSpy.mock.results).toMatchSnapshot();
  });

  it('Parses processingConfig correctly', async () => {
    AWSMock.setSDKInstance(AWS);

    AWSMock.mock('EKS', 'describeCluster', (params, callback) => {
      callback(null, mockCluster);
    });

    const createStateMachineSpy = jest.fn(

      // eslint-disable-next-line consistent-return
      (stateMachineObject) => (_.cloneDeepWith(JSON.parse(stateMachineObject.definition), (o) => {
        if (_.isObject(o) && o.image) {
          return {
            ...o,
            image: 'MOCK_IMAGE_PATH',
          };
        }
      })),
    );

    AWSMock.mock('StepFunctions', 'createStateMachine', (params, callback) => {
      createStateMachineSpy(params);
      callback(null, { stateMachineArn: 'test-machine' });
    });

    AWSMock.mock('StepFunctions', 'startExecution', (params, callback) => {
      callback(null, { executionArn: 'test-machine' });
    });

    const getItemSpy = jest.fn((x) => x);
    AWSMock.mock('DynamoDB', 'getItem', (params, callback) => {
      getItemSpy(params);
      callback(null, MockProcessingConfig);
    });

    await createPipeline('testExperimentId', processingConfigUpdate);
    expect(createStateMachineSpy.mock.results).toMatchSnapshot();
  });

  it('Pipeline is updated instead of created if an error is thrown.', async () => {
    AWSMock.setSDKInstance(AWS);

    const describeClusterSpy = jest.fn((x) => x);
    AWSMock.mock('EKS', 'describeCluster', (params, callback) => {
      describeClusterSpy(params);
      callback(null, mockCluster);
    });

    const createStateMachineSpy = jest.fn((stateMachineObject) => _.omit(stateMachineObject, 'definition'));
    AWSMock.mock('StepFunctions', 'createStateMachine', (params, callback) => {
      createStateMachineSpy(params);
      callback({ code: 'StateMachineAlreadyExists' }, null);
    });

    const updateStateMachineSpy = jest.fn((stateMachineObject) => _.omit(stateMachineObject, 'definition'));
    AWSMock.mock('StepFunctions', 'updateStateMachine', (params, callback) => {
      updateStateMachineSpy(params);
      callback(null, { stateMachineArn: 'test-machine' });
    });

    const startExecutionSpy = jest.fn((x) => x);
    AWSMock.mock('StepFunctions', 'startExecution', (params, callback) => {
      startExecutionSpy(params);
      callback(null, { executionArn: 'test-execution' });
    });

    await createPipeline('testExperimentId', processingConfigUpdate);
    expect(describeClusterSpy).toMatchSnapshot();
    expect(createStateMachineSpy.mock.results).toMatchSnapshot();

    expect(updateStateMachineSpy).toHaveBeenCalled();
    expect(updateStateMachineSpy.mock.results).toMatchSnapshot();

    expect(startExecutionSpy).toHaveBeenCalled();
    expect(startExecutionSpy.mock.results).toMatchSnapshot();
  });
});
