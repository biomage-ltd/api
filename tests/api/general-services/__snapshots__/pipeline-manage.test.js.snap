// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for pipeline services Create pipeline works 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "biomage-test",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "name": "biomage-test",
      },
    },
  ],
}
`;

exports[`test for pipeline services Create pipeline works 2`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "loggingConfiguration": Object {
        "level": "OFF",
      },
      "name": "biomage-pipeline-test-af3b5c3acaa6348165d8e5e73c70d93b732bf6db",
      "roleArn": "arn:aws:iam::test-account-id:role/state-machine-role-test",
      "tags": Array [
        Object {
          "key": "experimentId",
          "value": "testExperimentId",
        },
        Object {
          "key": "clusterEnv",
          "value": "test",
        },
        Object {
          "key": "sandboxId",
          "value": undefined,
        },
      ],
      "type": "STANDARD",
    },
  },
]
`;

exports[`test for pipeline services Create pipeline works 3`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "input": "{\\"samples\\":[\\"single-branch-in-map-state\\"]}",
      "stateMachineArn": "test-machine",
      "traceHeader": undefined,
    },
  },
]
`;

exports[`test for pipeline services Parses processingConfig correctly 1`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "Comment": "Pipeline for clusterEnv 'test'",
      "StartAt": "DeleteCompletedPipelineWorker",
      "States": Object {
        "ConfigureEmbedding": Object {
          "Catch": Array [
            Object {
              "ErrorEquals": Array [
                "EKS.409",
              ],
              "Next": "EndOfPipeline",
              "ResultPath": "$.error-info",
            },
          ],
          "Comment": "Attempts to create a Kubernetes Job for the pipeline server. Will swallow a 409 (already exists) error.",
          "End": true,
          "Parameters": Object {
            "CertificateAuthority": "AAAAAAAAAAA",
            "ClusterName": "biomage-test",
            "Endpoint": "https://test-endpoint.me/fgh",
            "Job": Object {
              "apiVersion": "batch/v1",
              "kind": "Job",
              "metadata": Object {
                "labels": Object {
                  "experimentId": "testExperimentId",
                  "taskName": "configureEmbedding",
                  "type": "pipeline",
                },
                "name": "remoter-client-4b7804f411921d45a04341ed534c4b1835a84206",
              },
              "spec": Object {
                "template": Object {
                  "metadata": Object {
                    "labels": Object {
                      "type": "pipeline",
                    },
                    "name": "remoter-client-testExperimentId",
                  },
                  "spec": Object {
                    "containers": Array [
                      Object {
                        "args": Array [
                          "{\\"experimentId\\":\\"testExperimentId\\",\\"taskName\\":\\"configureEmbedding\\",\\"config\\":{},\\"server\\":\\"remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local\\"}",
                        ],
                        "image": "MOCK_IMAGE_PATH",
                        "name": "remoter-client",
                      },
                    ],
                    "restartPolicy": "Never",
                  },
                },
              },
            },
            "LogOptions": Object {
              "RetrieveLogs": true,
            },
            "Namespace": "pipeline-test-namespace",
          },
          "Resource": "arn:aws:states:::eks:runJob.sync",
          "ResultPath": null,
          "Retry": Array [
            Object {
              "BackoffRate": 2,
              "ErrorEquals": Array [
                "EKS.409",
              ],
              "IntervalSeconds": 1,
              "MaxAttempts": 2,
            },
          ],
          "Type": "Task",
        },
        "DataIntegration": Object {
          "Catch": Array [
            Object {
              "ErrorEquals": Array [
                "EKS.409",
              ],
              "Next": "ConfigureEmbedding",
              "ResultPath": "$.error-info",
            },
          ],
          "Comment": "Attempts to create a Kubernetes Job for the pipeline server. Will swallow a 409 (already exists) error.",
          "Next": "ConfigureEmbedding",
          "Parameters": Object {
            "CertificateAuthority": "AAAAAAAAAAA",
            "ClusterName": "biomage-test",
            "Endpoint": "https://test-endpoint.me/fgh",
            "Job": Object {
              "apiVersion": "batch/v1",
              "kind": "Job",
              "metadata": Object {
                "labels": Object {
                  "experimentId": "testExperimentId",
                  "taskName": "dataIntegration",
                  "type": "pipeline",
                },
                "name": "remoter-client-7cdeb6ecdeb0e8c46825e54c41a8c132be4ae5a6",
              },
              "spec": Object {
                "template": Object {
                  "metadata": Object {
                    "labels": Object {
                      "type": "pipeline",
                    },
                    "name": "remoter-client-testExperimentId",
                  },
                  "spec": Object {
                    "containers": Array [
                      Object {
                        "args": Array [
                          "{\\"experimentId\\":\\"testExperimentId\\",\\"taskName\\":\\"dataIntegration\\",\\"config\\":{},\\"server\\":\\"remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local\\"}",
                        ],
                        "image": "MOCK_IMAGE_PATH",
                        "name": "remoter-client",
                      },
                    ],
                    "restartPolicy": "Never",
                  },
                },
              },
            },
            "LogOptions": Object {
              "RetrieveLogs": true,
            },
            "Namespace": "pipeline-test-namespace",
          },
          "Resource": "arn:aws:states:::eks:runJob.sync",
          "ResultPath": null,
          "Retry": Array [
            Object {
              "BackoffRate": 2,
              "ErrorEquals": Array [
                "EKS.409",
              ],
              "IntervalSeconds": 1,
              "MaxAttempts": 2,
            },
          ],
          "Type": "Task",
        },
        "DeleteCompletedPipelineWorker": Object {
          "Comment": "Deletes the previous server pipeline HelmRelease (Service+Job).",
          "Next": "LaunchNewPipelineWorker",
          "Parameters": Object {
            "CertificateAuthority": "AAAAAAAAAAA",
            "ClusterName": "biomage-test",
            "Endpoint": "https://test-endpoint.me/fgh",
            "Method": "DELETE",
            "Path": "/apis/helm.fluxcd.io/v1/namespaces/pipeline-test-namespace/helmreleases",
            "QueryParameters": Object {
              "labelSelector": Array [
                "type=pipeline",
              ],
            },
          },
          "Resource": "arn:aws:states:::eks:call",
          "ResultPath": null,
          "Type": "Task",
        },
        "EndOfPipeline": Object {
          "End": true,
          "Type": "Pass",
        },
        "Filters": Object {
          "ItemsPath": "$.samples",
          "Iterator": Object {
            "StartAt": "CellSizeDistributionFilter",
            "States": Object {
              "CellSizeDistributionFilter": Object {
                "Catch": Array [
                  Object {
                    "ErrorEquals": Array [
                      "EKS.409",
                    ],
                    "Next": "MitochondrialContentFilter",
                    "ResultPath": "$.error-info",
                  },
                ],
                "Comment": "Attempts to create a Kubernetes Job for the pipeline server. Will swallow a 409 (already exists) error.",
                "Next": "MitochondrialContentFilter",
                "Parameters": Object {
                  "CertificateAuthority": "AAAAAAAAAAA",
                  "ClusterName": "biomage-test",
                  "Endpoint": "https://test-endpoint.me/fgh",
                  "Job": Object {
                    "apiVersion": "batch/v1",
                    "kind": "Job",
                    "metadata": Object {
                      "labels": Object {
                        "experimentId": "testExperimentId",
                        "taskName": "cellSizeDistribution",
                        "type": "pipeline",
                      },
                      "name": "remoter-client-358a69da12f00cf47ef5507a34f40d5145df2052",
                    },
                    "spec": Object {
                      "template": Object {
                        "metadata": Object {
                          "labels": Object {
                            "type": "pipeline",
                          },
                          "name": "remoter-client-testExperimentId",
                        },
                        "spec": Object {
                          "containers": Array [
                            Object {
                              "args": Array [
                                "{\\"experimentId\\":\\"testExperimentId\\",\\"taskName\\":\\"cellSizeDistribution\\",\\"config\\":{},\\"server\\":\\"remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local\\"}",
                              ],
                              "image": "MOCK_IMAGE_PATH",
                              "name": "remoter-client",
                            },
                          ],
                          "restartPolicy": "Never",
                        },
                      },
                    },
                  },
                  "LogOptions": Object {
                    "RetrieveLogs": true,
                  },
                  "Namespace": "pipeline-test-namespace",
                },
                "Resource": "arn:aws:states:::eks:runJob.sync",
                "ResultPath": null,
                "Retry": Array [
                  Object {
                    "BackoffRate": 2,
                    "ErrorEquals": Array [
                      "EKS.409",
                    ],
                    "IntervalSeconds": 1,
                    "MaxAttempts": 2,
                  },
                ],
                "Type": "Task",
              },
              "ClassifierFilter": Object {
                "Catch": Array [
                  Object {
                    "ErrorEquals": Array [
                      "EKS.409",
                    ],
                    "Next": "NumGenesVsNumUmisFilter",
                    "ResultPath": "$.error-info",
                  },
                ],
                "Comment": "Attempts to create a Kubernetes Job for the pipeline server. Will swallow a 409 (already exists) error.",
                "Next": "NumGenesVsNumUmisFilter",
                "Parameters": Object {
                  "CertificateAuthority": "AAAAAAAAAAA",
                  "ClusterName": "biomage-test",
                  "Endpoint": "https://test-endpoint.me/fgh",
                  "Job": Object {
                    "apiVersion": "batch/v1",
                    "kind": "Job",
                    "metadata": Object {
                      "labels": Object {
                        "experimentId": "testExperimentId",
                        "taskName": "classifier",
                        "type": "pipeline",
                      },
                      "name": "remoter-client-6060600fe603377e418db0937ff8545ada4f2b52",
                    },
                    "spec": Object {
                      "template": Object {
                        "metadata": Object {
                          "labels": Object {
                            "type": "pipeline",
                          },
                          "name": "remoter-client-testExperimentId",
                        },
                        "spec": Object {
                          "containers": Array [
                            Object {
                              "args": Array [
                                "{\\"experimentId\\":\\"testExperimentId\\",\\"taskName\\":\\"classifier\\",\\"config\\":{},\\"server\\":\\"remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local\\"}",
                              ],
                              "image": "MOCK_IMAGE_PATH",
                              "name": "remoter-client",
                            },
                          ],
                          "restartPolicy": "Never",
                        },
                      },
                    },
                  },
                  "LogOptions": Object {
                    "RetrieveLogs": true,
                  },
                  "Namespace": "pipeline-test-namespace",
                },
                "Resource": "arn:aws:states:::eks:runJob.sync",
                "ResultPath": null,
                "Retry": Array [
                  Object {
                    "BackoffRate": 2,
                    "ErrorEquals": Array [
                      "EKS.409",
                    ],
                    "IntervalSeconds": 1,
                    "MaxAttempts": 2,
                  },
                ],
                "Type": "Task",
              },
              "DoubletScoresFilter": Object {
                "Catch": Array [
                  Object {
                    "ErrorEquals": Array [
                      "EKS.409",
                    ],
                    "Next": "EndOfMap",
                    "ResultPath": "$.error-info",
                  },
                ],
                "Comment": "Attempts to create a Kubernetes Job for the pipeline server. Will swallow a 409 (already exists) error.",
                "End": true,
                "Parameters": Object {
                  "CertificateAuthority": "AAAAAAAAAAA",
                  "ClusterName": "biomage-test",
                  "Endpoint": "https://test-endpoint.me/fgh",
                  "Job": Object {
                    "apiVersion": "batch/v1",
                    "kind": "Job",
                    "metadata": Object {
                      "labels": Object {
                        "experimentId": "testExperimentId",
                        "taskName": "doubletScores",
                        "type": "pipeline",
                      },
                      "name": "remoter-client-bff7a29bbbf49cb3e075d2f2e2e1873c9916e45f",
                    },
                    "spec": Object {
                      "template": Object {
                        "metadata": Object {
                          "labels": Object {
                            "type": "pipeline",
                          },
                          "name": "remoter-client-testExperimentId",
                        },
                        "spec": Object {
                          "containers": Array [
                            Object {
                              "args": Array [
                                "{\\"experimentId\\":\\"testExperimentId\\",\\"taskName\\":\\"doubletScores\\",\\"config\\":{\\"filterSettings\\":{\\"oneSetting\\":1},\\"oneSample\\":{\\"filterSettings\\":{\\"oneSetting\\":7}},\\"otherSample\\":{\\"filterSettings\\":{\\"oneSetting\\":15}}},\\"server\\":\\"remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local\\"}",
                              ],
                              "image": "MOCK_IMAGE_PATH",
                              "name": "remoter-client",
                            },
                          ],
                          "restartPolicy": "Never",
                        },
                      },
                    },
                  },
                  "LogOptions": Object {
                    "RetrieveLogs": true,
                  },
                  "Namespace": "pipeline-test-namespace",
                },
                "Resource": "arn:aws:states:::eks:runJob.sync",
                "ResultPath": null,
                "Retry": Array [
                  Object {
                    "BackoffRate": 2,
                    "ErrorEquals": Array [
                      "EKS.409",
                    ],
                    "IntervalSeconds": 1,
                    "MaxAttempts": 2,
                  },
                ],
                "Type": "Task",
              },
              "EndOfMap": Object {
                "End": true,
                "Type": "Pass",
              },
              "MitochondrialContentFilter": Object {
                "Catch": Array [
                  Object {
                    "ErrorEquals": Array [
                      "EKS.409",
                    ],
                    "Next": "ClassifierFilter",
                    "ResultPath": "$.error-info",
                  },
                ],
                "Comment": "Attempts to create a Kubernetes Job for the pipeline server. Will swallow a 409 (already exists) error.",
                "Next": "ClassifierFilter",
                "Parameters": Object {
                  "CertificateAuthority": "AAAAAAAAAAA",
                  "ClusterName": "biomage-test",
                  "Endpoint": "https://test-endpoint.me/fgh",
                  "Job": Object {
                    "apiVersion": "batch/v1",
                    "kind": "Job",
                    "metadata": Object {
                      "labels": Object {
                        "experimentId": "testExperimentId",
                        "taskName": "mitochondrialContent",
                        "type": "pipeline",
                      },
                      "name": "remoter-client-6ad152dff60ab5c322ab9eab3355b61c4fbca98f",
                    },
                    "spec": Object {
                      "template": Object {
                        "metadata": Object {
                          "labels": Object {
                            "type": "pipeline",
                          },
                          "name": "remoter-client-testExperimentId",
                        },
                        "spec": Object {
                          "containers": Array [
                            Object {
                              "args": Array [
                                "{\\"experimentId\\":\\"testExperimentId\\",\\"taskName\\":\\"mitochondrialContent\\",\\"config\\":{},\\"server\\":\\"remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local\\"}",
                              ],
                              "image": "MOCK_IMAGE_PATH",
                              "name": "remoter-client",
                            },
                          ],
                          "restartPolicy": "Never",
                        },
                      },
                    },
                  },
                  "LogOptions": Object {
                    "RetrieveLogs": true,
                  },
                  "Namespace": "pipeline-test-namespace",
                },
                "Resource": "arn:aws:states:::eks:runJob.sync",
                "ResultPath": null,
                "Retry": Array [
                  Object {
                    "BackoffRate": 2,
                    "ErrorEquals": Array [
                      "EKS.409",
                    ],
                    "IntervalSeconds": 1,
                    "MaxAttempts": 2,
                  },
                ],
                "Type": "Task",
              },
              "NumGenesVsNumUmisFilter": Object {
                "Catch": Array [
                  Object {
                    "ErrorEquals": Array [
                      "EKS.409",
                    ],
                    "Next": "DoubletScoresFilter",
                    "ResultPath": "$.error-info",
                  },
                ],
                "Comment": "Attempts to create a Kubernetes Job for the pipeline server. Will swallow a 409 (already exists) error.",
                "Next": "DoubletScoresFilter",
                "Parameters": Object {
                  "CertificateAuthority": "AAAAAAAAAAA",
                  "ClusterName": "biomage-test",
                  "Endpoint": "https://test-endpoint.me/fgh",
                  "Job": Object {
                    "apiVersion": "batch/v1",
                    "kind": "Job",
                    "metadata": Object {
                      "labels": Object {
                        "experimentId": "testExperimentId",
                        "taskName": "numGenesVsNumUmis",
                        "type": "pipeline",
                      },
                      "name": "remoter-client-28237f2e44fed034368b1ce69033029f9c089081",
                    },
                    "spec": Object {
                      "template": Object {
                        "metadata": Object {
                          "labels": Object {
                            "type": "pipeline",
                          },
                          "name": "remoter-client-testExperimentId",
                        },
                        "spec": Object {
                          "containers": Array [
                            Object {
                              "args": Array [
                                "{\\"experimentId\\":\\"testExperimentId\\",\\"taskName\\":\\"numGenesVsNumUmis\\",\\"config\\":{},\\"server\\":\\"remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local\\"}",
                              ],
                              "image": "MOCK_IMAGE_PATH",
                              "name": "remoter-client",
                            },
                          ],
                          "restartPolicy": "Never",
                        },
                      },
                    },
                  },
                  "LogOptions": Object {
                    "RetrieveLogs": true,
                  },
                  "Namespace": "pipeline-test-namespace",
                },
                "Resource": "arn:aws:states:::eks:runJob.sync",
                "ResultPath": null,
                "Retry": Array [
                  Object {
                    "BackoffRate": 2,
                    "ErrorEquals": Array [
                      "EKS.409",
                    ],
                    "IntervalSeconds": 1,
                    "MaxAttempts": 2,
                  },
                ],
                "Type": "Task",
              },
            },
          },
          "Next": "DataIntegration",
          "Type": "Map",
        },
        "LaunchNewPipelineWorker": Object {
          "Catch": Array [
            Object {
              "ErrorEquals": Array [
                "EKS.409",
              ],
              "Next": "Filters",
              "ResultPath": "$.error-info",
            },
          ],
          "Comment": "Attempts to create a Kubernetes Job+Service for the pipeline server. Will swallow a 409 (already exists) error.",
          "Next": "Filters",
          "Parameters": Object {
            "CertificateAuthority": "AAAAAAAAAAA",
            "ClusterName": "biomage-test",
            "Endpoint": "https://test-endpoint.me/fgh",
            "Method": "POST",
            "Path": "/apis/helm.fluxcd.io/v1/namespaces/pipeline-test-namespace/helmreleases",
            "RequestBody": Object {
              "apiVersion": "helm.fluxcd.io/v1",
              "kind": "HelmRelease",
              "metadata": Object {
                "annotations": Object {
                  "fluxcd.io/automated": "true",
                },
                "labels": Object {
                  "type": "pipeline",
                },
                "name": "remoter-server-testExperimentId",
                "namespace": "pipeline-test-namespace",
              },
              "spec": Object {
                "chart": Object {
                  "git": "git@github.com:biomage-ltd/pipeline",
                  "path": "remoter-server/chart",
                  "ref": "debug-jira-650",
                },
                "releaseName": "remoter-server-testExperimentId",
                "values": Object {
                  "awsAccountId": "test-account-id",
                  "awsRegion": "eu-west-1",
                  "clusterEnv": "test",
                  "experimentId": "testExperimentId",
                  "image": "MOCK_IMAGE_PATH",
                  "namespace": "pipeline-test-namespace",
                },
              },
            },
          },
          "Resource": "arn:aws:states:::eks:call",
          "ResultPath": null,
          "Retry": Array [
            Object {
              "BackoffRate": 2,
              "ErrorEquals": Array [
                "EKS.409",
              ],
              "IntervalSeconds": 1,
              "MaxAttempts": 2,
            },
          ],
          "Type": "Task",
        },
      },
    },
  },
]
`;

exports[`test for pipeline services Pipeline is updated instead of created if an error is thrown. 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "biomage-test",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "name": "biomage-test",
      },
    },
  ],
}
`;

exports[`test for pipeline services Pipeline is updated instead of created if an error is thrown. 2`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "loggingConfiguration": Object {
        "level": "OFF",
      },
      "name": "biomage-pipeline-test-af3b5c3acaa6348165d8e5e73c70d93b732bf6db",
      "roleArn": "arn:aws:iam::test-account-id:role/state-machine-role-test",
      "tags": Array [
        Object {
          "key": "experimentId",
          "value": "testExperimentId",
        },
        Object {
          "key": "clusterEnv",
          "value": "test",
        },
        Object {
          "key": "sandboxId",
          "value": undefined,
        },
      ],
      "type": "STANDARD",
    },
  },
]
`;

exports[`test for pipeline services Pipeline is updated instead of created if an error is thrown. 3`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "roleArn": "arn:aws:iam::test-account-id:role/state-machine-role-test",
      "stateMachineArn": "arn:aws:states:eu-west-1:test-account-id:stateMachine:biomage-pipeline-test-af3b5c3acaa6348165d8e5e73c70d93b732bf6db",
    },
  },
]
`;

exports[`test for pipeline services Pipeline is updated instead of created if an error is thrown. 4`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "input": "{\\"samples\\":[\\"single-branch-in-map-state\\"]}",
      "stateMachineArn": "arn:aws:states:eu-west-1:test-account-id:stateMachine:biomage-pipeline-test-af3b5c3acaa6348165d8e5e73c70d93b732bf6db",
      "traceHeader": undefined,
    },
  },
]
`;
